src/main/java/com/microsoft/azure/relay/HybridHttpConnection.java

```diff
@@ -324,13 +327,20 @@ public void setWriteTimeout(Duration writeTimeout) {
 
 				// When there is no request message body
 				if (this.writeBufferStream != null && this.writeBufferStream.position() > 0) {
-					return CompletableFuture.allOf(sendResponseTask, this.connection
-							.sendBytesOverRendezvousAsync(this.writeBufferStream, timeout).thenRun(() -> {
-								this.writeBufferStream.clear();
-								if (this.writeBufferFlushTimer != null) {
-									this.writeBufferFlushTimer.cancel();
-								}
-							}));
+					return sendResponseTask.thenCompose($void -> {
+						// Get a new buffer backed by the non empty segment of the write buffer array
+						this.writeBufferStream.flip();
+						ByteBuffer trimmedBuffer = ByteBuffer.wrap(
+								this.writeBufferStream.array(), 
+								this.writeBufferStream.position(), 
+								this.writeBufferStream.remaining());
```

Comment:
Can you help me understand this? Why is a clone of the writeBufferStream needed (the wrap which I believe creates something basically equivalent)?

Reply:
It's to wrap only the part of the buffer that has been written to, so the empty bytes in the buffer aren't sent.

---
In reply to: [292073271](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292073271) [](ancestors = 292073271)

Reply:
Does that actually happen without the wrap here? 

---
In reply to: [292098393](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292098393) [](ancestors = 292098393,292073271)

Reply:
You are right, what I needed was the flip() on the buffer, not wrapping it.

---
In reply to: [292098796](https://github.com/Azure/azure-relay-java/pull/45#discussion_r292098796) [](ancestors = 292098796,292098393,292073271)